templates/components/multiagent/typescript/workflow/tools.ts

```diff
@@ -0,0 +1,324 @@
+  for (const toolCall of toolCalls) {
+    const tool = tools.find((tool) => tool.metadata.name === toolCall.name);
+    if (!tool) {
+      throw new Error(`Tool ${toolCall.name} not found`);
+    }
+    const toolMsg = await callSingleTool(toolCall, tool, (msg: string) => {
+      ctx.writeEventToStream(
+        new AgentRunEvent({
+          name: agentName,
+          text: msg,
+          type: "progress",
+          data: {
+            id: progressId,
+            total: totalSteps,
+            current: currentStep,
+          },
+        }),
+      );
+      currentStep++;
+    });
+    toolMsgs.push(toolMsg as ChatMessage);
+  }
+  return toolMsgs;
+};
+
+export const callSingleTool = async (
+  toolCall: ToolCall,
+  tool: BaseToolWithCall,
+  eventEmitter: (msg: string) => void,
+) => {
+  if (eventEmitter) {
+    eventEmitter(
+      `Calling tool ${toolCall.name} with input: ${JSON.stringify(toolCall.input)}`,
+    );
+  }
+
+  const toolOutput = await callTool(tool, toolCall, {
+    log: () => {},
+    error: (...args: unknown[]) => {
+      console.error(`Tool ${toolCall.name} got error:`, ...args);
+      if (eventEmitter) {
+        eventEmitter(`Tool ${toolCall.name} got error: ${args.join(" ")}`);
+      }
+      return {
+        content: JSON.stringify({
+          error: args.join(" "),
+        }),
+        role: "user",
+        options: {
+          toolResult: {
+            id: toolCall.id,
+            result: JSON.stringify({
+              error: args.join(" "),
+            }),
+            isError: true,
+          },
+        },
+      };
+    },
+    warn: () => {},
+  });
+
+  return {
+    content: JSON.stringify(toolOutput.output),
+    role: "user",
+    options: {
+      toolResult: {
+        result: toolOutput.output,
+        isError: toolOutput.isError,
+        id: toolCall.id,
+      },
+    },
+  };
+};
+
+class ChatWithToolsResponse {
+  toolCalls: ToolCall[];
+  toolCallMessage?: ChatMessage;
+  responseGenerator?: AsyncGenerator<ChatMessage>;
+
+  constructor(options: {
+    toolCalls: ToolCall[];
+    toolCallMessage?: ChatMessage;
+    responseGenerator?: AsyncGenerator<ChatMessage>;
+  }) {
+    this.toolCalls = options.toolCalls;
+    this.toolCallMessage = options.toolCallMessage;
+    this.responseGenerator = options.responseGenerator;
+  }
+
+  isCallingDifferentTools() {
+    // toolCalls is have different tool names
+    const toolNames = this.toolCalls.map((toolCall) => toolCall.name);
+    const uniqueToolNames = new Set(toolNames);
+    return uniqueToolNames.size > 1;
+  }
+
+  isCallingTool() {
+    return this.toolCalls.length > 0;
+  }
+
+  toolName() {
+    return this.toolCalls[0].name;
+  }
+
+  async asFullResponse() {
+    if (!this.responseGenerator) {
+      throw new Error("No response generator");
+    }
+    let fullResponse = "";
+    for await (const chunk of this.responseGenerator) {
+      fullResponse += chunk.content;
+    }
+    return fullResponse;
```

Comment:
just realizing this, when we need this, we always want the message, right? maybe returning then this?
```suggestion
    return {
          role: "assistant",
          content: fullResponse,
        };
```